<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />
    <title>Asynchronous JavaScript</title>
  </head>
  <body>
    <h1>Asynchronous JavaScript</h1>
    <script>
      //   const second = () => {
      //     setTimeout(() => {
      //       console.log("Second");
      //     }, 2000); // timer
      //   };
      //   const first = () => {
      //     console.log("hey there"); //1
      //     second(); //function call in function, 3
      //     console.log("The End"); //2
      //   };
      //   first();

      //   function getRecipe() {
      //     setTimeout(() => {
      //       const recipeID = [523, 883, 432, 974];
      //       console.log(recipeID);

      //       setTimeout(
      //         (id) => {
      //           const recipe = {
      //             title: "Fresh tomato pasta",
      //             publisher: "Jonas",
      //           };
      //           console.log(`${id}: ${recipe.title}`);
      //           setTimeout(
      //             (publisher) => {
      //               const recipe = { title: "Italian Pizza", publisher: "Jonas" };
      //               console.log(recipe);
      //             },
      //             1500,
      //             recipe.publisher
      //           );
      //         },
      //         1000,
      //         recipeID[2]
      //       );
      //     }, 1500);
      //   }
      //   getRecipe();
      //   promise를 사용할 때 인자를 2개 받는데, resolve는 promise 작동 성공, reject는 실패
      const getIDs = new Promise((resolve, reject) => {
        setTimeout(() => {
          resolve([523, 883, 432, 974]);
        }, 1500);
      });
      const getRecipe = (recID) => {
        return new Promise((resolve, reject) => {
          setTimeout(
            (id) => {
              const recipe = { title: "Italian Pizza", publisher: "Jonas" };
              resolve(`${id}: ${recipe.title}`);
            },
            1500,
            recID
          );
        });
      };
      const getRelated = (publisher) => {
        return new Promise((resolve, reject) => {
          setTimeout(
            (pub) => {
              const recipe = { title: "Italian Pizza", publisher: "Jonas" };
              resolve(`${pub}: ${recipe.title}`);
            },
            1500,
            publisher
          );
        });
      };
      //   getIDs
      //     .then((IDs) => {
      //       console.log(IDs);
      //       return getRecipe(IDs[2]);
      //     })
      //     .then((recipe) => {
      //       console.log(recipe);
      //       //   return getRelated(recipe.publisher);
      //       return getRelated(`Jonas`);
      //     })
      //     .then((recipe) => {
      //       console.log(recipe);
      //     })
      //     .catch((error) => {
      //       //   console.log(error);
      //       console.log("Error Happened");
      //     });
      async function getRecipesAW() {
        const IDs = await getIDs;
        console.log(IDs);
        const recipe = await getRecipe(IDs[2]);
        console.log(recipe);
        const related = await getRelated("Jonas");
        console.log(related);
        return recipe;
      }
      getRecipesAW().then((result) => console.log(`${result}`));
      function getWeather(woeid) {
        fetch(
          `https://crossorigin.me/https://www.metaweather.com/api/location/${woeid}/`
        )
          .then((result) => {
            // console.log(result);
            return result.json();
          })
          .then((data) => {
            console.log(data);
            const today = data.consolidated_weather[0];
            console.log(
              `Temperatures in ${data.title} stay between ${today.min_temp} and ${today.max_temp}`
            );
          })
          .catch((error) => console.log(error));
      }
      getWeather(2487956);
      getWeather(44418);

      async function getRecipesAW(woeid) {
        try {
          const result = await fetch(
            `https://crossorigin.me/https://www.metaweather.com/api/location/${woeid}/`
          );
          const data = await result.json();
          const tomorrow = data.consolidated_weather[1];
          console.log(
            `Temperatures in ${data.title} stay between ${tomorrow.min_temp} and ${tomorrow.max_temp}`
          );
          console.log(data);
          return data;
        } catch (error) {
          console.log(error);
        }
      }
      getWeather(2487956);
      let dataLondon;
      getWeather(44418).then((data) => {
        dataLondon = data;
        console.log(dataLondon);
      });
    </script>
  </body>
</html>
